### PR TITLE
Facebook Local Server

### PR DESCRIPTION
**Pull request by derekchen14**
Prepare web back-end in Flask to properly serve Fb-neg task and front-end templates to display user interface.

### CODE REVIEW

File: fb-negotiation/web/main/backend.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+from core.event import Event
+# from analysis.analyze_strategy import StrategyAnalyzer
+
+class DatabaseManager(BaseDatabaseManager):
+    @classmethod
+    def add_survey_table(cls, cursor):
+        cursor.execute(
+            '''CREATE TABLE survey (name text, chat_id text, partner_type text, fluent integer,
+            honest integer, persuasive integer, fair integer, negotiator integer, coherent integer, comments text)''')
+
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py]: only keep negotiator and comments

File: fb-negotiation/web/main/backend.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                                  "survey".format(chat_id, user_name, userid, agent_idx))
+                self.end_chat_and_finish(cursor, userid, message=msg)
+
+        if game_over:
+            if not self.is_user_partner_bot(cursor, userid):
+                verify_chat(partner_id, 1 - agent_idx, True)
+            verify_chat(userid, agent_idx, False)
+            return True
+
+        return False
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py]: no need to verify. cleanup these functions.

File: fb-negotiation/web/main/backend.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+        _, game_complete = self.is_game_over(userid)
+        if game_complete:
+            msg = self.messages.ChatCompleted
+            partner_msg = msg
+        else:
+            msg = self.messages.ChatIncomplete
+            partner_msg = msg
+
+        return msg, partner_msg
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py]: completion messages need to be updated.

File: fb-negotiation/web/main/backend.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                cursor = self.conn.cursor()
+                u = self._get_user_info_unchecked(cursor, userid)
+                self._update_user(cursor, userid, connected_status=1)
+                self.send(userid, Event.OfferEvent(u.agent_index,
+                                                   offer,
+                                                   str(time.time())))
+        except sqlite3.IntegrityError:
+            print("WARNING: Rolled back transaction")
+            return None
+
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py:96]: remove make_offer.

File: fb-negotiation/web/main/backend.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            with self.conn:
+                cursor = self.conn.cursor()
+                u = self._get_user_info_unchecked(cursor, userid)
+                self._update_user(cursor, userid, connected_status=1)
+                self.send(userid, Event.QuitEvent(u.agent_index,
+                                                  None,
+                                                  str(time.time())))
+        except sqlite3.IntegrityError:
+            print("WARNING: Rolled back transaction")
+            return None
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py]: remove quit

File: fb-negotiation/web/main/backend.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+            with self.conn:
+                cursor = self.conn.cursor()
+                user_info = self._get_user_info_unchecked(cursor, userid)
+                cursor.execute('''SELECT scenario_id FROM chat WHERE chat_id=?''', (user_info.chat_id,))
+                scenario_id = cursor.fetchone()[0]
+                _update_scenario_db(user_info.chat_id, scenario_id, user_info.partner_type)
+                cursor.execute('INSERT INTO survey VALUES (?,?,?,?,?,?,?,?,?,?)',
+                               (userid, user_info.chat_id, user_info.partner_type,
+                                data['fluent'], data['honest'], data['persuasive'],
+                                data['fair'], data['negotiator'], data['coherent'], data['comments']))
```
Comment (hhexiy) [inline fb-negotiation/web/main/backend.py]: update values. only need negotiator.

File: fb-negotiation/web/main/db_reader.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import math
+import json
+
+from cocoa.web.main.db_reader import DatabaseReader as BaseDatabaseReader
+from cocoa.core.util import write_json
+
```
Comment (hhexiy) [inline fb-negotiation/web/main/db_reader.py:6]: this module needs to be updated to fit fb-negotiation data structure.

**Review by hhexiy — COMMENTED** [review]
Looks good! Main thing is to implement the "dump" functions and clean up.

File: fb-negotiation/web/main/backend.py (hunk 1 above)
Comment (derekchen14) [reply fb-negotiation/web/main/backend.py]: it seems items like chat_id is also necessary for bookkeeping?  Also, how does negotiator differ from partner-type?  I imagine one of them specifies whether that dialogue was human or bot.  In any case, I removed the majority of other columns.

**Pull request merged**